staging/src/k8s.io/apiserver/pkg/storage/cacher/cache_watcher.go

```diff
@@ -195,10 +195,14 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
 		//     the bookmarkAfterResourceVersion we want. We do that to make progress
 		//     as clients can re-establish a new watch with the given RV and receive
 		//     further notifications.
+		// (b) We have sent the bookmark or the bookmark hasn't been requested.
+		//     In those cases we rely on the higher-level component (i.e., the watch server)
+		//     To eventually stop the watcher, for example when the client disconnected
+		//     or when the request times out.
 		graceful := func() bool {
 			c.stateMutex.Lock()
 			defer c.stateMutex.Unlock()
-			return c.state == cacheWatcherBookmarkReceived
+			return c.state != cacheWatcherWaitingForBookmark
```

Comment:
the question is whether we would like to drain the buffer in cases when:

- we have sent the special bookmark event to the client
- the special bookmark hasn't been requested (a standard watch request)

note that graceful termination will also allow for sending the initial events (implicitly)

Reply:
Neither of those cases are actually clear to me:

1) when we have sent special bookmark, the client is already synced is just slow with processing - graceful shutdown doesn't buy as much..
2) if the bookmark wasn't requested, the situation is similar (slow watcher) and again it's unclear if we get much from it

Do you have any scenarios on your mind where it would visibly help?